File: TimeDilationScript.m
%% Script developed by Anum Pirkani
% For details, contact: anum.apirkani@gmail.com

clc;
clear; 
close all;

%% Initial Variables
   SpeedLight         = 299792458;
   SpeedMovingTarget1 = 0.8 * SpeedLight;    
   SpeedMovingTarget2 = 0.99 * SpeedLight;   
   LorentzFactor1     = 1 / sqrt(1 - (SpeedMovingTarget1^2 / SpeedLight^2));
   LorentzFactor2     = 1 / sqrt(1 - (SpeedMovingTarget2^2 / SpeedLight^2));

   VelocityRatioValues = linspace(0, 1, 600);
   LorentzFactorValues = 1 ./ sqrt(1 - VelocityRatioValues.^2);

   fig = figure('Name', 'Time Dilation', 'NumberTitle', 'off',...
                'Position', [50, 50, 1500, 600], 'Color', [0.1, 0.1, 0.1]);

   videoFile           = VideoWriter('TimeDilation.mp4', 'MPEG-4');
   videoFile.FrameRate = 30;
   open(videoFile);

%% Lorentz Factor vs Velocity
   subplot(131);
   hold on;
   FirstPlot = plot(NaN, NaN,'m','LineWidth',5);
   xlabel('Velocity Ratio (v/c)','FontSize',12,'FontWeight','bold','Color','white');
   ylabel('Time Dilation (γ)','FontSize',12,'FontWeight','bold','Color','white');
   grid on;
   grid minor;
   xlim([0 1]);
   ylim([0.9 7]);
   set(gca,'Color',[0.2 0.2 0.2],'XColor','white','YColor','white');

%% Time Progress
   subplot(132);
   hold on;
   StationaryPlot = plot(0, 0, 'b', 'LineWidth', 5);
   MovingPlot1    = plot(0, 0, 'r', 'LineWidth', 5);
   MovingPlot2    = plot(0, 0, 'g', 'LineWidth', 5); 
   xlabel('Stationary Time (s)', 'FontSize', 12, 'FontWeight', 'bold', 'Color', 'white');
   ylabel('Elapsed Time (s)', 'FontSize', 12, 'FontWeight', 'bold', 'Color', 'white');
   legend({'Stationary Time', 'Moving: 0.8*c', 'Moving: 0.99*c'}, 'TextColor', 'k', 'Location', 'northwest');
   grid on;
   grid minor;
   xlim([0 60]);
   ylim([0 60]);
   set(gca,'Color',[0.2 0.2 0.2],'XColor','white','YColor','white');

%% Clocks
   subplot(133);
   hold on;
   axis equal;
   xlim([-2 2]);
   ylim([-1.5 2]);
   set(gca,'XColor','none','YColor','none');

   ThetaClock = linspace(0, 2*pi, 100);
   xCircle    = cos(ThetaClock);
   yCircle    = sin(ThetaClock);

   % Stationary Clock
     fill(0+0.4*xCircle, 1.2+0.4*yCircle, 'b', 'LineWidth', 3);
     h1 = line([0, 0], [1.2, 1.2 + 0.3], 'Color', 'k', 'LineWidth', 4);
     text(0, 1.8, 'Stationary Clock', 'HorizontalAlignment', 'center', 'FontSize', 12, 'FontWeight', 'bold', 'Color', 'k');
     TextStationary = text(0, 0.6, '0.00 s', 'HorizontalAlignment', 'center', 'FontSize', 14, 'FontWeight', 'bold', 'Color', 'k');

   % Moving Clock: 0.8*c
     fill(-0.8+0.4*xCircle, -0.6+0.4*yCircle, 'r', 'LineWidth', 3);
     h2 = line([-0.8, -0.8], [-0.6, -0.6 + 0.3], 'Color', 'k', 'LineWidth', 4);
     text(-0.8, 0.05, 'Moving: 0.8*c', 'HorizontalAlignment', 'center', 'FontSize', 12, 'FontWeight', 'bold', 'Color', 'k');
     TextMoving1 = text(-0.8, -1.2, '0.00 s', 'HorizontalAlignment', 'center', 'FontSize', 14, 'FontWeight', 'bold', 'Color', 'k');

   % Moving Clock: 0.99*c
     fill(0.8+0.4*xCircle, -0.6+0.4*yCircle, 'g', 'LineWidth', 3);
     h3 = line([0.8, 0.8], [-0.6, -0.6 + 0.3], 'Color', 'k', 'LineWidth', 4);
     text(0.8, 0.05, 'Moving: 0.99*c', 'HorizontalAlignment', 'center', 'FontSize', 12, 'FontWeight', 'bold', 'Color', 'k');
     TextMoving2 = text(0.8, -1.2, '0.00 s', 'HorizontalAlignment', 'center', 'FontSize', 14, 'FontWeight', 'bold', 'Color', 'k');

%% Animation
   InitialTime          = 0;
   MaximumTime          = 60;
   dt                   = 0.1;
   StationaryTimeValues = 0;
   MovingTimeValues1    = 0;
   MovingTimeValues2    = 0;
   i = 1;

   tic;
   while InitialTime < MaximumTime
         InitialTime = i * dt;
         if toc < InitialTime
            continue; 
         end
    
         TimeMoving1 = InitialTime / LorentzFactor1;
         TimeMoving2 = InitialTime / LorentzFactor2;

         set(h1, 'XData', [0, 0 + 0.3*cos(-InitialTime * pi/30 + pi/2)], ...
                 'YData', [1.2, 1.2 + 0.3*sin(-InitialTime * pi/30 + pi/2)]);
         set(TextStationary, 'String', sprintf('%.2fs', InitialTime));

         set(h2, 'XData', [-0.8, -0.8 + 0.3*cos(-TimeMoving1 * pi/30 + pi/2)], ...
                 'YData', [-0.6, -0.6 + 0.3*sin(-TimeMoving1 * pi/30 + pi/2)]);
         set(TextMoving1, 'String', sprintf('%.2fs', TimeMoving1));

         set(h3, 'XData', [0.8, 0.8 + 0.3*cos(-TimeMoving2 * pi/30 + pi/2)], ...
                 'YData', [-0.6, -0.6 + 0.3*sin(-TimeMoving2 * pi/30 + pi/2)]);
         set(TextMoving2, 'String', sprintf('%.2fs', TimeMoving2));

         StationaryTimeValues = [StationaryTimeValues, InitialTime]; %#ok<*AGROW>
         MovingTimeValues1    = [MovingTimeValues1, TimeMoving1];
         MovingTimeValues2    = [MovingTimeValues2, TimeMoving2];
         set(StationaryPlot, 'XData', StationaryTimeValues, 'YData', StationaryTimeValues);
         set(MovingPlot1, 'XData', StationaryTimeValues, 'YData', MovingTimeValues1);
         set(MovingPlot2, 'XData', StationaryTimeValues, 'YData', MovingTimeValues2);
   
         if i > length(VelocityRatioValues)
            break;
         end
         set(FirstPlot, 'XData', VelocityRatioValues(1:i), 'YData', LorentzFactorValues(1:i)); 
    
         drawnow;
         
         frame = getframe(fig);
         writeVideo(videoFile, frame);
    
         i = i + 1;
   end
   
 close(videoFile);

File: TimeDilationScript.py
# Script developed by Anum Pirkani
# For details, please contact anum.apirkani@gmail.com

import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation

# Constants
SpeedLight = 299792458
SpeedMovingTarget1 = 0.8 * SpeedLight
SpeedMovingTarget2 = 0.99 * SpeedLight
LorentzFactor1 = 1 / np.sqrt(1 - (SpeedMovingTarget1 ** 2 / SpeedLight ** 2))
LorentzFactor2 = 1 / np.sqrt(1 - (SpeedMovingTarget2 ** 2 / SpeedLight ** 2))

VelocityRatioValues = np.linspace(0, 1, 600)
LorentzFactorValues = 1 / np.sqrt(1 - VelocityRatioValues**2)

fig, axes = plt.subplots(1, 3, figsize=(15, 6), facecolor=(0.1, 0.1, 0.1))
plt.subplots_adjust(wspace=0.3)

# Lorentz Factor vs. Velocity
ax1 = axes[0]
ax1.set_xlim(0, 1)
ax1.set_ylim(0.9, 7)
ax1.set_xlabel("Velocity Ratio (v/c)", fontsize=12, fontweight="bold", color="white")
ax1.set_ylabel("Time Dilation (γ)", fontsize=12, fontweight="bold", color="white")
ax1.grid(True, linestyle="--", alpha=0.6)
ax1.set_facecolor((0.2, 0.2, 0.2))
ax1.tick_params(colors="white")
lorentz_plot, = ax1.plot([], [], "m", linewidth=5)

# Time Progress
ax2 = axes[1]
ax2.set_xlim(0, 60)
ax2.set_ylim(0, 60)
ax2.set_xlabel("Stationary Time (s)", fontsize=12, fontweight="bold", color="white")
ax2.set_ylabel("Elapsed Time (s)", fontsize=12, fontweight="bold", color="white")
ax2.grid(True, linestyle="--", alpha=0.6)
ax2.set_facecolor((0.2, 0.2, 0.2))
ax2.tick_params(colors="white")
stationary_plot, = ax2.plot([], [], "b", linewidth=5, label="Stationary Time")
moving_plot1, = ax2.plot([], [], "r", linewidth=5, label="Moving: 0.8c")
moving_plot2, = ax2.plot([], [], "g", linewidth=5, label="Moving: 0.99c")
ax2.legend(loc="upper left", fontsize=10, facecolor="white", edgecolor="white")

# Clock
ax3 = axes[2]
ax3.set_xlim(-2, 2)
ax3.set_ylim(-1.5, 2)
ax3.set_xticks([])
ax3.set_yticks([])
ax3.set_facecolor((0.2, 0.2, 0.2))

theta    = np.linspace(0, 2 * np.pi, 100)
x_circle = 0.4 * np.cos(theta)
y_circle = 0.4 * np.sin(theta)

ax3.fill(0 + x_circle, 1.2 + y_circle, "b")
ax3.fill(-0.8 + x_circle, -0.6 + y_circle, "r")
ax3.fill(0.8 + x_circle, -0.6 + y_circle, "g")

h1, = ax3.plot([0, 0], [1.2, 1.5], "k", linewidth=4)
h2, = ax3.plot([-0.8, -0.8], [-0.6, -0.3], "k", linewidth=4)
h3, = ax3.plot([0.8, 0.8], [-0.6, -0.3], "k", linewidth=4)

ax3.text(0, 1.8, "Stationary Clock", ha="center", fontsize=12, color="white")
ax3.text(-0.8, 0.1, "Moving: 0.8*c", ha="center", fontsize=12, color="white")
ax3.text(0.8, 0.1, "Moving: 0.99*c", ha="center", fontsize=12, color="white")

text_stationary = ax3.text(0, 0.6, "0.00 s", ha="center", fontsize=12, color="white")
text_moving1 = ax3.text(-0.8, -1.2, "0.00 s", ha="center", fontsize=12, color="white")
text_moving2 = ax3.text(0.8, -1.2, "0.00 s", ha="center", fontsize=12, color="white")

stationary_time = []
moving_time1 = []
moving_time2 = []

def update(frame):
    time_stationary = frame * 0.1
    time_moving1 = time_stationary / LorentzFactor1
    time_moving2 = time_stationary / LorentzFactor2

    # Update Lorentz Factor Graph
    lorentz_plot.set_data(VelocityRatioValues[:frame], LorentzFactorValues[:frame])

    # Update Time Progress Graph
    stationary_time.append(time_stationary)
    moving_time1.append(time_moving1)
    moving_time2.append(time_moving2)

    stationary_plot.set_data(stationary_time, stationary_time)
    moving_plot1.set_data(stationary_time, moving_time1)
    moving_plot2.set_data(stationary_time, moving_time2)

    # Update Clock Hands
    h1.set_data([0, 0.3 * np.cos(-time_stationary * np.pi / 30 + np.pi / 2)], 
                [1.2, 1.2 + 0.3 * np.sin(-time_stationary * np.pi / 30 + np.pi / 2)])

    h2.set_data([-0.8, -0.8 + 0.3 * np.cos(-time_moving1 * np.pi / 30 + np.pi / 2)], 
                [-0.6, -0.6 + 0.3 * np.sin(-time_moving1 * np.pi / 30 + np.pi / 2)])

    h3.set_data([0.8, 0.8 + 0.3 * np.cos(-time_moving2 * np.pi / 30 + np.pi / 2)], 
                [-0.6, -0.6 + 0.3 * np.sin(-time_moving2 * np.pi / 30 + np.pi / 2)])

    # Update Time Text
    text_stationary.set_text(f"{time_stationary:.2f} s")
    text_moving1.set_text(f"{time_moving1:.2f} s")
    text_moving2.set_text(f"{time_moving2:.2f} s")

    return lorentz_plot, stationary_plot, moving_plot1, moving_plot2, text_stationary, text_moving1, text_moving2

anim = animation.FuncAnimation(fig, update, frames=600, interval=30, blit=True)

anim.save("TimeDilationPython.gif", writer="pillow", fps=30)

plt.show()
